### ISSUE TITLE
Markers of "end of text" appear when generating samples

### ISSUE THREAD

**Issue by studiosi** [openai/gpt-2#35]
When generating samples interactively sometimes an "end of text" marker is included on the text. Is this related to the sanitization of the training data?

**Comment by WuTheFWasThat** [comment]
it's due to how our model was trained.  (if you wanted, the sampling code could be modified to either reject endoftext while sampling, or to simply end the sample)

**Comment by studiosi** [comment]
Correct.

**Comment by summerstay** [comment]
How would I modify the sampling code to reject <|end of text|> tokens?

**Comment by gwern** [comment]
You could simply post-process it to drop everything after a `<|endoftext|>`. For example, split on '===' and pipe each sample through `sed -e '/<|endoftext|>/{s/<|endoftext|>.*//;q}'`, which quits on EOT, truncating the sample and dropping everything after that.

**Comment by summerstay** [comment]
Sure, but it would be better if there was some way to tell it to pick a less likely token and keep going, so that we don't get a disappointingly short generation.

**Comment by WuTheFWasThat** [comment]
you can change the code to have the logit for the endoftext token (50256 i think) to negative infinity so they never get sampled

**Comment by summerstay** [comment]
In the definition for _top_k, I see the line
` tf.ones_like(logits, dtype=logits.dtype) * -1e10,`
could I make that modification to the logit for the endof text token here? I'm confused about how exactly to do it.

**Comment by loganriggs** [comment]
This works in npshepperd's fork. Shouldn't be too different for openai's version.

    try:
        scalar_of_logit = tf.reshape(logits[:,50256], [])
        condition = tf.less(scalar_of_logit, 0)
        scalar = tf.cond(
            condition,
            lambda: np.inf,
            lambda: -1*np.inf,
        )
        logits = tf.concat([logits[:,:50256], logits[:,50256, tf.newaxis]*scalar], axis=1)
    except:
        pass
    samples = tf.multinomial(logits, num_samples=1, output_dtype=tf.int32)

You may need to import numpy as np in that file. 

A short explanation: We use a try-except block to pass through the initialization phase. We check if the logit is negative (iirc it will if it is less than %50 in the softmax). We then reassign logits through a concatentation because reassigning by splicing is horrible for tf v1. We left the "samples = " line so you should be able to identify where this code should be placed.

Note: you could change 50256 to whatever token you'd like, but be sure to add the last section of logits like:
`            logits = tf.concat(first_part, logits[:,N, tf.newaxis]*scalar, last_part], axis=1)
`
